flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,279 @@
+        this.numRequestedBuffers = new AtomicInteger(0);
+        this.bufferReclaimRequestListeners = new ArrayList<>();
+        this.isInitialized = false;
+    }
+
+    @Override
+    public void setup(BufferPool bufferPool, List<TieredStorageMemorySpec> storageMemorySpecs) {
+        this.bufferPool = bufferPool;
+        for (TieredStorageMemorySpec memorySpec : storageMemorySpecs) {
+            checkState(
+                    !tieredMemorySpecs.containsKey(memorySpec.getOwner()),
+                    "Duplicated memory spec.");
+            tieredMemorySpecs.put(memorySpec.getOwner(), memorySpec);
+            numTotalGuaranteedBuffers += memorySpec.getNumGuaranteedBuffers();
+        }
+
+        if (needPeriodicalCheckReclaimBuffer) {
+            this.executor =
+                    Executors.newSingleThreadScheduledExecutor(
+                            new ThreadFactoryBuilder()
+                                    .setNameFormat("buffer reclaim checker")
+                                    .setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE)
+                                    .build());
+        }
+
+        this.isInitialized = true;
+    }
+
+    @Override
+    public void listenBufferReclaimRequest(Runnable onBufferReclaimRequest) {
+        bufferReclaimRequestListeners.add(onBufferReclaimRequest);
+    }
+
+    /**
+     * Request a {@link BufferBuilder} instance from {@link BufferPool} for a specific owner. The
+     * {@link TieredStorageMemoryManagerImpl} will not check whether a buffer can be requested and
+     * only record the total number of requested buffers. If the buffers in the {@link BufferPool}
+     * is not enough, this will request each tiered storage to reclaim the buffers as much as
+     * possible.
+     *
+     * <p>Note that synchronization is not necessary for this method, as it can only be called by
+     * the task thread.
+     *
+     * @return the requested buffer
+     */
+    @Override
+    public BufferBuilder requestBufferBlocking() {
+        checkIsInitialized();
+
+        tryReclaimBuffers();
+        CompletableFuture<MemorySegment> requestedBuffer =
+                CompletableFuture.supplyAsync(
+                        () -> {
+                            try {
+                                return bufferPool.requestMemorySegmentBlocking();
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
+                            }
+                        });
+
+        ScheduledFuture<?> scheduledFuture = getPeriodicalCheckReclaimBufferFuture();
+
+        MemorySegment memorySegment = null;
+        try {
+            memorySegment = requestedBuffer.get();
+        } catch (InterruptedException | ExecutionException e) {
+            ExceptionUtils.rethrow(e);
+        }
+
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(false);
```

Comment:
Instead of canceling the scheduled future, I'd suggest to check whether the request buffer future is completed in the periodical task.

Reply:
Fixed by a `CompletableFuture<Void>`. And these two lines are removed.